File: src/Java/algorithms/datastructure/ValidSudoku.java
package algorithms.datastructure; 

import java.util.*; 

public class ValidSudoku {
    /* Function: Check if a sudoku is valid till now. The empty part is '.'. 
     * Solution: Use three matrixes as hash table. 
     */ 
    
    public boolean isValidSudoku(char[][] board) {
        if (board == null || board.length != 9 || board[0].length != 9) {
            return false; 
        }
        
        boolean[][][] check = new boolean[3][9][9]; // row, col, grid
        
        for (int i = 0; i < board.length; ++i) {
            for (int j = 0; j < board[0].length; ++j) {
                if (board[i][j] == '.') {
                    continue; 
                }
                
                if (!valid(i, j, board[i][j] - '0', check)) {
                    return false; 
                }
            }
        }
        
        return true; 
    }
    
    private boolean valid(int x, int y, int val, boolean[][][] check) {
        if (val <= 0 || val > 9) {
            return false; 
        }
        
        if (check[0][x][val - 1]) {
            return false; 
        } else {
            check[0][x][val - 1] = true; 
        }
        
        if (check[1][y][val - 1]) {
            return false; 
        } else {
            check[1][y][val - 1] = true;
        }
        
        if (check[2][x / 3 * 3 + y / 3][val - 1]) {
            return false; 
        } else {
            check[2][x / 3 * 3 + y / 3][val - 1] = true; 
        }
        
        return true; 
    }
} 

File: src/CPP/src/leetcode/ValidSudoku.cpp
/*
 * [Source] https://leetcode.com/problems/valid-sudoku/
 * [Difficulty]: Easy
 * [Tag]: Hash Table
 */

#include <iostream>

using namespace std;

// [Solution]: For each row and column and square, hash exist number
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/datastructure/ValidSudoku.java
 */

/* Java solution
public class Solution {
    public boolean isValidSudoku(char[][] board) {
        int i, j, pos, id;
        int[][] allConstrains = new int[3 * 9][9]; 
        for (i = 0; i < 9; ++i) { 
            for (j = 0; j < 9; ++j) {
                if (board[i][j] == '.') {
                    continue;
                }
                pos = board[i][j] - '1'; 
                if (allConstrains[i][pos] == 1) { // Check row. 
                    return false; 
                } else {
                    allConstrains[i][pos] = 1;
                }
                if (allConstrains[9 + j][pos] == 1) { // Check col. 
                    return false; 
                } else {
                    allConstrains[9 + j][pos] = 1;
                }
                id = getBlockId(i, j); 
                if (allConstrains[id][pos] == 1) { // Check block. 
                    return false; 
                } else {
                    allConstrains[id][pos] = 1; 
                }
            }
        }
        return true; 
    }
    
    private int getBlockId(int x, int y) {
        int xAxis = x / 3; 
        int yAxis = y / 3; 
        return (18 + 3 * xAxis + yAxis); 
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
